File: lib/fluxsource/testpatternfluxsource.cc
#include "lib/core/globals.h"
#include "lib/data/fluxmap.h"
#include "lib/fluxsource/fluxsource.h"
#include "lib/fluxsource/fluxsource.pb.h"

class TestPatternFluxSource : public TrivialFluxSource
{
public:
    TestPatternFluxSource(const TestPatternFluxSourceProto& config):
        _config(config)
    {
        _extraConfig.mutable_drive()->set_tracks("c0-255h0-1");
    }

    ~TestPatternFluxSource() {}

public:
    std::unique_ptr<const Fluxmap> readSingleFlux(int track, int side) override
    {
        auto fluxmap = std::make_unique<Fluxmap>();

        while (fluxmap->duration() < (_config.sequence_length_ms() * 1e6))
        {
            fluxmap->appendInterval(
                _config.interval_us() * (double)TICKS_PER_US);
            fluxmap->appendPulse();
        }

        return fluxmap;
    }

    void recalibrate() override {}

private:
    const TestPatternFluxSourceProto& _config;
};

std::unique_ptr<FluxSource> FluxSource::createTestPatternFluxSource(
    const TestPatternFluxSourceProto& config)
{
    return std::make_unique<TestPatternFluxSource>(config);
}

File: java/com/cowlark/fluxengine/fluxsource/TestPatternFluxSource.java
package com.cowlark.fluxengine.fluxsource;

import com.cowlark.fluxengine.config.ConfigBuilder;
import com.cowlark.fluxengine.config.ConfigProto;
import com.cowlark.fluxengine.data.Fluxmap;
import com.cowlark.fluxengine.wiring.FluxEngine;

/**
 * A flux source which generates a synthetic test pattern, ported from
 * lib/fluxsource/testpatternfluxsource.cc.
 */
public class TestPatternFluxSource extends TrivialFluxSource
{
    private final double intervalUs;
    private final double sequenceLengthMs;
    protected ConfigProto extraConfig;

    public TestPatternFluxSource(TestPatternFluxSourceProto config)
    {
        intervalUs = config.getIntervalUs();
        sequenceLengthMs = config.getSequenceLengthMs();

        ConfigProto.Builder builder = ConfigProto.newBuilder();
        builder.getDriveBuilder().setTracks("c0-255h0-1");
        extraConfig = builder.build();
    }

    @Override
    public void adjustConfig(ConfigBuilder configBuilder)
    {
        configBuilder.mergeConfig(extraConfig);
    }

    @Override
    public Fluxmap readSingleFlux(FluxReadParameters parameters)
    {
        Fluxmap fluxmap = new Fluxmap();

        while (fluxmap.durationNs() < (sequenceLengthMs * 1e6))
        {
            fluxmap.appendInterval((int) (intervalUs * FluxEngine.TICKS_PER_US));
            fluxmap.appendPulse();
        }

        return fluxmap;
    }
}